### PR TITLE
Use Isolate to Quantize

### PR DESCRIPTION
**Pull request by deep-gaurav**
Current computation method is very heavy on older devices.
This shifts computation to an isolate using compute.
this is only done is filters are null, since i could not find a way to pass filters/functions as message to isolate

### CODE REVIEW

**Comment by deep-gaurav** [comment]
I signed it!

File: packages/palette_generator/lib/palette_generator.dart [hunk 1]
```diff
@@ -1134,15 +1182,19 @@ class _ColorCutQuantizer {
...
   }
 
-  List<PaletteColor> _quantizePixels(
-    int maxColors,
-    Map<Color, int> histogram,
+
+  static List<PaletteColor> _quantizePixels(
+    Map args,List<PaletteFilter> filters
   ) {
+    print("Isolate started");
```
Comment (gspencergoog) [inline packages/palette_generator/lib/palette_generator.dart]: Remove debugging output.

File: packages/palette_generator/lib/palette_generator.dart [hunk 2]
```diff
@@ -1106,8 +1106,56 @@ class _ColorCutQuantizer {
 
     final ByteData imageData =
         await image.toByteData(format: ui.ImageByteFormat.rawRgba);
+    if(filters==null)
+      return await compute(_quantizefromByte, {
+        "byteData":imageData,
```
Comment (gspencergoog) [inline packages/palette_generator/lib/palette_generator.dart]: String literals should all be single-quote strings.

File: packages/palette_generator/lib/palette_generator.dart [hunk 3]
```diff
@@ -1106,8 +1106,56 @@ class _ColorCutQuantizer {
...
+    List<PaletteColor> _paletteColors =map["paletteColors"];
+    
+
+    const int quantizeWordWidth = 5;
+    const int quantizeChannelWidth = 8;
+    const int quantizeShift = quantizeChannelWidth - quantizeWordWidth;
+    const int quantizeWordMask =
+        ((1 << quantizeWordWidth) - 1) << quantizeShift;
+
+    Color quantizeColor(Color color) {
```
Comment (gspencergoog) [inline packages/palette_generator/lib/palette_generator.dart]: Don't duplicate code: make `quantizeColor` a `static`, private function.

You can move the `quantizeWordMask` definition (and dependent values) inside of the function.

**Comment by gspencergoog** [comment]
Also, fix the analyzer errors and warnings.

File: packages/palette_generator/lib/palette_generator.dart (hunk 1 above)
Comment (deep-gaurav) [reply packages/palette_generator/lib/palette_generator.dart]: Removed, thanks

File: packages/palette_generator/lib/palette_generator.dart (hunk 2 above)
Comment (deep-gaurav) [reply packages/palette_generator/lib/palette_generator.dart]: Done, thanks

File: packages/palette_generator/lib/palette_generator.dart [hunk 4]
```diff
@@ -1106,8 +1106,56 @@ class _ColorCutQuantizer {
...
+
+  static List<PaletteColor> _quantizefromByte(Map map){
+    ByteData bd = map["byteData"];
+    int maxColors = map["maxColors"];
+    int width = map["width"];
+    int height = map["height"];
+    Rect region = map["region"];
+    List<PaletteFilter> filters = map["filters"];
+    List<PaletteColor> _paletteColors =map["paletteColors"];
+
```
[the remark this answers is not in the document]
Comment (deep-gaurav) [reply packages/palette_generator/lib/palette_generator.dart]: compute only lets one parameter to pass so I used map to pass the values, I'm not exactly sure how to pass multiple parameters, besides using a list

[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/palette_generator/lib/palette_generator.dart]: No, I mean remove the extra blank line here.  There are two, when usually only one is used.

**Comment by gspencergoog** [comment]
Thanks for the updates @deep-gaurav!  Let me know when you're done fixing the analyzer problems, and I'll take another look.

**Comment by goderbauer** [comment]
@deep-gaurav Do you still plan to come back to this to fix the analyzer issues?

**Comment by goderbauer** [comment]
I am closing this since there has been no follow-up.

This approach is really interesting though. If somebody has the time for it, this is worth to be picked up again.

**Pull request closed without merging**